src/switchroot/ostree-mount-util.h

```diff
@@ -106,6 +120,96 @@ read_proc_cmdline_ostree (void)
   return ret;
 }
 
+static inline int
+append_latest_link(char* dir_str, int* len, int* cap) {
+  char* append_from = dir_str + *len;
+  int appended_len = 0;
+  DIR __attribute__ ((cleanup(close_dir))) *dir = opendir(dir_str);
+  if (!dir)
+    {
+      fprintf(stderr, "'%s' %d", strerror(errno), __LINE__);
+      return 1;
+    }
+
+  struct dirent *ent = 0;
+  for (time_t latest = 0; (ent = readdir(dir)); )
+    {
+      if (ent->d_type != DT_LNK)
+        continue;
+
+      struct stat lsb;
+      lsb.st_mtime = 0;
+      char ent_str[PATH_MAX + sizeof (ent->d_name)];
+      snprintf(ent_str, sizeof (ent_str), "%s/%s", dir_str, ent->d_name);
+      if (lstat(ent_str, &lsb) == -1)
+        {
+          fprintf(stderr, "'%s' %d", strerror(errno), __LINE__);
+          return 1;
+        }
+
+      if (latest <= lsb.st_mtime)
+        {
+          appended_len = snprintf(append_from, *cap, "/%s", ent->d_name);
+          latest = lsb.st_mtime;
+        }
+    }
+
+  *len += appended_len;
+  *cap = PATH_MAX - *len;
+
+  return 0;
+}
```

Comment:
So, as I take it, this method searches `$prefix/ostree/` and appends the latest symlink.

Maybe it would be good to only look at `$prefix/ostree/boot*` (perhaps even limited to all possible combinations) ?

Otherwise creating a link in there (even if you shouldn't do so) seems to have potential to break any attempt to boot.

Sorting by modification timestamps seems a bit fragile to me, can't that be done by simply looking at the number in the link ?

Reply:
> So the current logic simply picks the latest, as defined by modification timestamp, file from `/sysroot/ostree/boot.$x/$stateroot/$bootcsum/$checkoutnum` as checkout to use ?

Yes.

> 
> I think that will lead to unexpected results at best, and unexpected vulnerabilities at worst, and I guess that's why this is a Draft currently.

Yes again, I'm just trying things at the moment.

> 
> Basing the decision on the _modified timestamp_ could have the very slim chance of actually booting an older, but later modified, deployment, I guess that situation would be rare.

Yes I am worried about this too, expect the implementation to change.

> 
> It would also mean, that you can boot a (signed) kernel with a different checkout than it was deployed with, which could open up new vulnerabilities, e.g. new deployment relies on kernel feature / initramfs changes that aren't present.
> 
> But the `$bootcsum` should be computable by the loader (or maybe even the kernel/initramfs itself), so while you can't embed the hash inside the images (which is the data being hashed), perhaps it could be passed by the UKI-stub or loader. Then you wouldn't need to "guess" it, and can ensure that a deployment would only ever be booted by a image it was build with. (Side-Note: That could even work if the bootcsum ends up being the hash of the UKI, but in that case it could be preferable to use the PECOFF hash that would be signed).

I am using a very limited bootloader that boots either an A or B boot partition, that contains an Android Boot Image. But there is very limited data available in the partition that I can be able to use in order to identify if I should use the latest or the latest-1 version. If that makes sense.

UKI's probably have much more options.

> 
> Rollbacks / multiple deployments, seem to be impossible with this logic, since even with an older image the latest checkout is used. I think the information of _which deployment_ to use comes inherently from an unverifiable source and the best you can do is probably to limit the image+deployment combinations to ones where the image booted and the image of the deployment match (bootcsum). And I believe that is true even if you implement automatic health checks, since at some point you need to trust something, and realistically you want to give users the option of choosing to boot rollback image+deployment (e.g. by means of UEFI boot selection / sd-boot).

I only want Android style A/B rollbacks, nothing else.

> 
> In any case you're putting your trust in nobody tampering with the actual filesystem, which is quite a far shot by today's standards ...

Will need to be hardened.

I think you can take this PR with a pinch of salt at the minute, it will be drastically changed. But thanks for the early feedback.

Reply:
> I am using a very limited bootloader that boots either an A or B boot partition, that contains an Android Boot Image. But there is very limited data available in the GPT partition table that I can be able to use in order to identify if I should use the latest or the latest-1 version. If that makes sense.

Ah, that makes things a bit easier :thinking: 
I think it might end up being a very similar trade-off like you have to do with UKI, since the information on what to boot (A or B) comes from an untrusted / unverified source.

> > In any case you're putting your trust in nobody tampering with the actual filesystem, which is quite a far shot by today's standards ...
> Will need to be hardened.

It was more of a general comment, a lot of things currently depend on the filesystem not having been tampered with.

> I think you can take this PR with a pinch of salt at the minute, it will be drastically changed. But thanks for the early feedback.

It is very nice to see some attempts at implementing it, even if it wouldn't fit my use case in the end :rocket: 
As always, an actual (incomplete) implementation is much easier to argue about than some abstract idea.